### PR TITLE
Deprecate asns/:id

### PR DESCRIPTION
**Pull request by shamrickus**
## What does this PR (Pull Request) do?

- [x] This PR deprecates the `asns/:id` endpoint

## Which Traffic Control components are affected by this PR?

- Documentation
- Traffic Ops

## What is the best way to verify this PR?
Query `/asns/:id` and confirm deprecation notice is displayed.

## The following criteria are ALL met by this PR

- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

File: docs/source/api/v2/asns_id.rst [hunk 1]
```diff
@@ -23,6 +23,8 @@
 
 ``GET``
 =======
+.. deprecated:: ATCv4
```
Comment (mhoppa) [inline docs/source/api/v2/asns_id.rst]: deprecation notice should go in the v1 documentation and then this GET documentation should be deleted

File: traffic_ops/traffic_ops_golang/routing/routes.go [hunk 2]
```diff
@@ -150,7 +150,7 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
 
 		//ASN: CRUD
 		{api.Version{2, 0}, http.MethodGet, `asns/?$`, api.ReadHandler(&asn.TOASNV11{}), auth.PrivLevelReadOnly, Authenticated, nil, 273877722, noPerlBypass},
-		{api.Version{2, 0}, http.MethodGet, `asns/{id}$`, api.ReadHandler(&asn.TOASNV11{}), auth.PrivLevelReadOnly, Authenticated, nil, 223008984, noPerlBypass},
+		{api.Version{2, 0}, http.MethodGet, `asns/{id}$`, api.DeprecatedReadHandler(&asn.TOASNV11{}, util.StrPtr("GET /asns")), auth.PrivLevelReadOnly, Authenticated, nil, 223008984, noPerlBypass},
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/routing/routes.go]: this should be deleted from 2.0 support and the 1.x route should be changed to use DeprecatedReadHandler

File: traffic_ops/app/lib/API/Asn.pm [hunk 3]
```diff
@@ -73,7 +73,12 @@ sub show {
 	my $self = shift;
 	my $id   = $self->param('id');
 
+	my $alt = "GET /asns with the 'id' parameter";
+
 	my $rs_data = $self->db->resultset("Asn")->search( { 'me.id' => $id }, { prefetch => ['cachegroup'] } );
+	if ( !defined($rs_data) ) {
+		return $self->with_deprecation("Resource not found", "error", 404, $alt);
```
Comment (mhoppa) [inline traffic_ops/app/lib/API/Asn.pm]: woot! making perl better :D

File: traffic_control/clients/python/trafficops/tosession.py [hunk 4]
```diff
@@ -308,8 +308,8 @@ def create_asn(self, query_params=None):
 		:raises: Union[LoginError, OperationError]
 		"""
 
-	@api_request('get', 'asns/{asn_id:d}', ('2.0',))
-	def get_asn_by_id(self, asn_id=None):
+	@api_request('get', 'asns', ('2.0',))
+	def get_asn_by_id(self, asn_id=None, query_params=None):
```
Comment (ocket8888) [inline traffic_control/clients/python/trafficops/tosession.py]: There's already a `get_asns` that does `GET /asns` with optional query parameters; this method should get deleted instead.

**Comment by shamrickus** [comment]
Not ready for merge, waiting on some changes from #4452

**Comment by shamrickus** [comment]
Ready for review again

File: traffic_control/clients/python/trafficops/tosession.py [hunk 5]
```diff
@@ -285,41 +285,6 @@ def create_asn(self, data=None):
...
-		Get ASN by ID
-		:ref:`to-api-asns-id`
-		:param asn_id: The ID of the ASN to retrieve
-		:type asn_id: int
-		:rtype: Tuple[Union[Dict[str, Any], List[Dict[str, Any]]], requests.Response]
-		:raises: Union[LoginError, OperationError]
-		"""
-
-	@api_request('put', 'asns/{asn_id:d}', ('2.0',))
 	def update_asn(self, asn_id=None, query_params=None):
```
Comment (ocket8888) [inline traffic_control/clients/python/trafficops/tosession.py]: `asn_id` should no longer be used as a param for `update_asn` - I don't think it causes any formatting problems, but it also doesn't do anything; you have to use `query_params` to pass the ASN's ID.

File: traffic_control/clients/python/trafficops/tosession.py [hunk 6]
```diff
@@ -330,8 +295,8 @@ def update_asn(self, asn_id=None, query_params=None):
 		:raises: Union[LoginError, OperationError]
 		"""
 
-	@api_request('delete', 'asns/{asn_id:d}', ('2.0',))
-	def delete_asn(self, asn_id=None):
+	@api_request('delete', 'asns', ('2.0',))
+	def delete_asn(self, asn_id=None, query_params=None):
```
Comment (ocket8888) [inline traffic_control/clients/python/trafficops/tosession.py]: Same as above

**Pull request merged**